### PR TITLE
Opening pull request to compare @wchargin changes

### PR DESCRIPTION
**Pull request by vsoch**

### CODE REVIEW

File: scripts/build_static_site.sh
Comment (vsoch) [inline scripts/build_static_site.sh]: I'm glad this could be removed and cleaned up.

File: scripts/build_static_site.sh
Comment (vsoch) [inline scripts/build_static_site.sh]: Just curious - what are your reasons for having preference for printf over echo?

File: scripts/entrypoint.sh
Comment (vsoch) [inline scripts/entrypoint.sh]: I'm not familiar with the syntax here, could you tell me the logic?

Comment (vsoch) [inline scripts/entrypoint.sh]: I'm not familiar with the syntax here, could you tell me the logic? E.g., what defaults to what when what is set, etc.

File: scripts/entrypoint.sh
Comment (vsoch) [inline scripts/entrypoint.sh]: Question - so we keep the definition of load_cmd here just to show the user but then run below in the subshell, correct? So keeping load_cmd is more of a courtesy to the user.

File: scripts/entrypoint.sh
Comment (vsoch) [inline scripts/entrypoint.sh]: This is something that I did but I'm just seeing now - is there ever a situation when the user would want to be using an update branch that already exists? E.g., so they would do just a checkout without -b, and then possibly a pull then push?

File: scripts/entrypoint.sh
Comment (vsoch) [inline scripts/entrypoint.sh]: Just curious, why not leave /bin/bash to explicitly show the shell being used?

File: scripts/pull_request.sh
Comment (vsoch) [inline scripts/pull_request.sh]: This header is really important because GitHub is deprecating GET params for the user token. This was never an issue here, but it's on my mind because I've been updating older scripts.

File: scripts/pull_request.sh
Comment (vsoch) [inline scripts/pull_request.sh]: what inputs args go here?

**Review by vsoch — COMMENTED** [review]
@wchargin the job ran and opened a PR so it looks good via actual testing, I have just a few questions I'm interested to ask then we should be good to merge into master.

**Comment by vsoch** [comment]
For future record:

 - [here](https://github.com/vsoch/sourcecred-cred/pull/9/checks?check_run_id=451561518) is the GitHub action
 - [here](https://github.com/vsoch/sourcecred-cred/pull/9) is the PR that triggered it

File: scripts/build_static_site.sh
Comment (wchargin) [reply scripts/build_static_site.sh]: From the commit message:

> Per POSIX, new code [should not use `echo`][echo]. The behavior is
> undefined when the input contains backslashes. This commit fixes `echo`s
> of non-constant strings.
> 
> [echo]: https://pubs.opengroup.org/onlinepubs/009695399/utilities/echo.html#tag_04_41_16

<https://github.com/vsoch/cred-action/pull/4/commits/2c5b595cdb77984fee623be25a746b015dddaafc>

File: scripts/entrypoint.sh
Comment (wchargin) [reply scripts/entrypoint.sh]: As mentioned in the commit message, this is called alternate value
expansion. You can get a good explanation of it reading `man bash` and
searching for “alternate value”:

> `${parameter:+word}`: **Use Alternate Value.**  If `parameter` is null
> or unset, nothing is substituted, otherwise the expansion of `word` is
> substituted.

The previous commit message mentions the motivation:

> Fix unquoted parameter expansions
>
> The statements patched in this commit behaved incorrectly when the
> parameters in question contained whitespace or globs.

Comment (wchargin) [reply scripts/entrypoint.sh]: The logic is the same as before. When `SC_WEIGHTS` is unset, it expands
to nothing. When `SC_WEIGHTS` is set, it expands to the two words
`--weights` and `"${GITHUB_WORKSPACE}/${SC_WEIGHTS}`", which (notably)
is quoted.

File: scripts/entrypoint.sh
Comment (wchargin) [reply scripts/entrypoint.sh]: Good catch; thanks! I meant to remove `load_cmd` along with `build_cmd`
in [`716df126e`][commit] but forgot. I’ll remove this so that they don’t
go out of sync.

[commit]: https://github.com/vsoch/cred-action/pull/4/commits/716df126e260680933fd36d7dd86cafd11ea69a5

File: scripts/entrypoint.sh
Comment (wchargin) [reply scripts/entrypoint.sh]: I’m not sure; I, too, was a bit surprised by the branch usage. I would
have expected this to just be committed onto a detached head and pushed
with `HEAD:refs/heads/update-"$(date -u +%s)"`.

File: scripts/entrypoint.sh
Comment (wchargin) [reply scripts/entrypoint.sh]: From the commit message:

> These scripts already declare their interpreter in the shebang;
> specifying a shell or options is generally discouraged because it pokes
> into the implementation details.

<https://github.com/vsoch/cred-action/pull/4/commits/87bbc83e6ae58185111392998a539e02660ffdb2>

To elaborate slightly: it would be valid to refactor the script in
question to use `/bin/sh` instead of `/bin/bash`, or vice versa, or to
use `csh` or something else, or to not expect to be run with `set -e` in
effect and therefore use other patterns around exit statuses. All of
these would break when explicitly running with `/bin/bash -e`.

File: scripts/pull_request.sh
Comment (wchargin) [reply scripts/pull_request.sh]: I’m not sure that I understand; did I port the header wrong? I was a bit
confused by the original `Accept: foo; bar` syntax; I’m used to seeing
comma-separated values, so I used those here.

File: scripts/pull_request.sh
Comment (wchargin) [reply scripts/pull_request.sh]: From the commit message:

> Shell scripts should invoke `main` as `main "$@"` so that the main
> function can actually read those arguments. We don’t use them now, but
> this kind of error can be pretty confusing to debug later in case we do
> want to use them.

<https://github.com/vsoch/cred-action/pull/4/commits/e153beb11d6b981a08efaa95f1eb636fc0372ca7>

**Review by wchargin — COMMENTED** [review]
Thanks for the review! I’ll make the change noted inline and take a look
at the latest CI run, then we can merge these with rebase.

File: scripts/build_static_site.sh
Comment (vsoch) [reply scripts/build_static_site.sh]: Looks like I need to better read your commit messages, they are beautifully documented.

File: scripts/entrypoint.sh
Comment (vsoch) [reply scripts/entrypoint.sh]: Thank you!

File: scripts/entrypoint.sh
Comment (vsoch) [reply scripts/entrypoint.sh]: okay, great! This is what I would want to do as well - it's hugely likely they would get out of sync, and the action will show the commands run without the print.

File: scripts/entrypoint.sh
Comment (vsoch) [reply scripts/entrypoint.sh]: I'd be good if you wanted to change this behavior. Right now we assume the branch doesn't exist - for the most part this will be the case, but as an example, @Beanow wants to use a cron that runs every 6 hours. He'd either need a more detailed timestamp, an automated merge and delete of the branch (instead of pull request) or a fix as you suggested.

File: scripts/pull_request.sh
Comment (vsoch) [reply scripts/pull_request.sh]: That's fair! Good to double check that the intention wasn't to actually have any input args (at least for now).

**Review by wchargin — APPROVED** [review]
Approving modulo re-running the tests:
<https://github.com/vsoch/sourcecred-cred/pull/9#issuecomment-588636751>

In the future it’d probably be best to let the author of the commits in
question open the appropriate pull requests, [as previously planned][1].
For instance, I would have preferred that the PR be on a branch other
than `wchargin-wip-changes`; branches like `*wip*` are generally
specifically intended to not be merged. It’s also weird that the
“author” of the PR is not the author of the actual code. We have a
“Paired-with” notion in SourceCred, but this isn’t quite that, because
the original code was written by just me, so your reviews should earn
cred for “real reviews of others’ code” rather than “self-reviews
responding to comments”. So I’m not sure what the right process is to
teach SourceCred the actual origin of this pull request; we might have
to invent something new.

But given that all the review comments and discussion are now in this
PR, perhaps it’s better to just merge this as is. Does that sound good
to you? If so, I will go ahead and “rebase and merge” this (rather than
squashing, which is our default) to preserve the commit history, as it
sounds like you agreed that the commit messages were helpful.

[1]: https://github.com/sourcecred/cred/pull/17#issuecomment-586600135

**Comment by vsoch** [comment]
I'd be good with that! But note that we can't do the final test, because the hourly rate limit goes over  - https://github.com/vsoch/sourcecred-cred/pull/9 I'm not sure what changed (and I'm not using the API really actively anywhere else) so I'm wondering if this is going to be a deal breaker for most? If it's merged and then the repository transferred to SourceCred, will the API limits be increased?

Let me know your thoughts, and I'm happy to do what you think is best moving forward. Also thank you for the feedback on discourse - I definitely messed this up and can do better next time.

File: scripts/build_static_site.sh
Comment (vsoch) [inline scripts/build_static_site.sh]: @wchargin is something happening so that the project file isn't actually honored? Based on [combined.json](https://github.com/vsoch/sourcecred-cred/blob/master/projects/combined.json), shouldn't it be the case that we don't run sourcecred over every repository in the org?

**Comment by vsoch** [comment]
Hey @wchargin we've just merged #6 and are ready to give this branch another go! When you have some time, would you care to resolve the conflicts? I can also offer to give it a shot, although I think you are likely the best one to do it. Let me know your preference. After that, I'll update the commit again in the vsoch/sourcecred-cred repository and give it a final test!

**Comment by wchargin** [comment]
Per discussion, @vsoch plans to write a local test harness for this
repository so that we can be more confident making changes to it, and
I’ll hold off on updating this base branch until then.

**Comment by vsoch** [comment]
@wchargin the test is added via #14, so you should be unblocked to rebase, and then have your changes tested. Thanks to @Beanow for the speedy review!

**Comment by wchargin** [comment]
Okay, I’ve rebased and resolved all the (many) conflicts. I couldn’t
find any way to run tests locally, but the GitHub CI workflow passed.
Is that sufficient, in your estimation? If so, please let me know, and I
will rebase and merge this pull request (rather than squashing it).

**Comment by vsoch** [comment]
@wchargin thank you for the rebase, I'm good with the changes and don't want to nit pick because this has been open for quite some time, and I think we should favor moving forward over perfection.

> I couldn’t find any way to run tests locally, but the GitHub CI workflow passed.

I'm hoping that if/when @Beanow does work to refactor the sourcecred command line util (and it works more like Jekyll) running locally should be much easier. For example, [urlchecker-action](https://github.com/urlstechie/urlchecker-action) serves as a wrapper for [urlchecker-python](https://github.com/urlstechie/urlchecker-python) so you can run the same check with one command line run that would be mapped into the action:

```bash
urlchecker check <options> .
```  
We will hopefully get there! The nice thing about the action is that while we aren't close to that, it provides a wrapper to the multiple load / scores commands so that it's easier for a GitHub org to deploy it.

> Is that sufficient, in your estimation? If so, please let me know, and I will rebase and merge this pull request (rather than squashing it).

The test implemented here is a base case - it has a very simple project file (just a few GitHub repos) and then the same for weights. I would say that yes, this test passing is sufficient. @Beanow would you care to review this as well? I'll bring up some other details that I think are outside of scope for this PR that might be commented on here:

 - I am not sure that we should have the pull request itself happen with the action - it feels like the action should generate the prototype and the user can then decide how to proceed. We can still make it easy by providing templates in example-github-workflows that show different use cases.
 - The cache will need to be added

@Beanow if you have a chance to review so @wchargin can rebase and we can finalize, it would be greatly appreciated! We should talk about the steps still required before transferring to sourceced, and then working on the github-example-workflows.

**Review by Beanow — COMMENTED** [review]
I'm only able to provide a gloss-over at the moment. I've read through the PR per-commit as they have important context. Agree it's better not to squash here.

The changes look good and are pretty uncontroversial improvements. :+1:
Added one nit.

---

As far as testing goes, I would say the test plan currently is:
 - The CI dry-run doesn't crash.
 - The CI dry-run's logged output "seems legit".

In that output I'm seeing the correct repositories were loaded (as opposed to say, the entire `@sourcecred` org when loading again for scores).

The static site build has the right project `PROJECT_IDS: ["@sourcecred"]`.

It's only a smoke test with manual high-level checks. But I would encourage merging based on that.

As @vsoch points out, some of these responsibilities that we currently use bash for, would be part of SourceCred's own CLI and validation tools with the planned instance system. It doesn't make sense to me to invest heavily in testing frameworks for individual scripts at this point with that in mind, but focus on the interface (e.g. triggering the GH action) first.

On top of that, I think `build_static_site.sh` could be slimmed down in the context of GH actions in follow ups. Because unlike what it was originally written for, GH actions don't need to "play nice" with the host machine. It owns it's own dedicated environment. So having to mirror the same data twice, because `build_static_site.sh` uses a neatly isolated temp dir and throws it away afterwards while we still needed this data, is not very efficient.

So with both of that done, we would have much less bash code left. The main contribution here is to set a great example of good shell code practices for the parts that remain.

File: scripts/pull_request.sh
Comment (wchargin) [reply scripts/pull_request.sh:8]: Good point; thanks. Done.

**Comment by wchargin** [comment]
Great, thanks.

Yes, running tests locally would be great. As is, to run tests on each
commit I have to push one of them to GitHub, wait for it to queue the
test, then push the next one, and so on. It would be much easier to just
run `git rebase origin/master --exec ./run_tests.sh`.

That said, each commit passed, so I’ll go ahead and rebase and merge.
Thanks for the reviews.

**Pull request merged**
